dev_docs/tutorials/testing_plugins.mdx

```diff
@@ -0,0 +1,1343 @@
+Use them whenever adding an integration test is hard and slow due to complex setup or the number of logic permutations.
+Since all external core and plugin dependencies are mocked, you don't have the guarantee that the whole system works as
+expected.
+Pros:
+- fast
+- easier to debug
+
+Cons:
+- doesn't test against real dependencies
+- doesn't cover integration with other plugins
+
+###### Example
+You can leverage existing unit-test infrastructure for this. You should add `*.test.ts` file and use dependencies mocks to cover the functionality with a broader test suit that covers:
+- input permutations
+- input edge cases
+- expected exception
+- interaction with dependencies
+```typescript
+// src/plugins/my_plugin/server/formatter.test.ts
+describe('TextFormatter', () => {
+  describe('format()', () => {
+    const sanitizer = sanitizerMock.createSetup();
+    sanitizer.sanitize.mockImplementation((input: string) => `sanitizer result:${input}`);
+
+    it('formats text to a ... format', async () => {
+      expect(await TextFormatter.format('aaa', sanitizer)).toBe('...');
+    });
+
+    it('calls Sanitizer.sanitize with correct arguments', async () => {
+      await TextFormatter.format('aaa', sanitizer);
+      expect(sanitizer.sanitize).toHaveBeenCalledTimes(1);
+      expect(sanitizer.sanitize).toHaveBeenCalledWith('aaa');
+    });
+
+    it('throws MisformedTextError if passed string contains banned symbols', async () => {
+      sanitizer.sanitize.mockRejectedValueOnce(new MisformedTextError());
+      await expect(TextFormatter.format('any', sanitizer)).rejects.toThrow(MisformedTextError);
+    });
+    // ... other tests
+  });
+});
+```
+
+#### Integration tests
+Depending on the number of external dependencies, you can consider implementing several high-level integration tests. 
+They would work as a set of [smoke tests](https://en.wikipedia.org/wiki/Smoke_testing_(software)) for the most important functionality.
+Main subjects for tests should be:
+- authenticated / unauthenticated access to an endpoint.
+- endpoint validation (params, query, body).
+- main business logic.
+- dependencies on other plugins.
+
+##### Functional Test Runner
+If your plugin relies on the elasticsearch server to store data and supports additional configuration, you can leverage the Functional Test Runner(FTR) to implement integration tests. 
+FTR bootstraps an elasticsearch and a Kibana instance and runs the test suite against it.
+Pros:
+- runs the whole Elastic stack
+- tests cross-plugin integration
+- emulates a real user interaction with the stack
+- allows adjusting config values
+
+Cons:
+- slow start
+- hard to debug
+- brittle tests
+
+###### Example
+You can reuse existing [api_integration](/test/api_integration/config.js) setup by registering a test file within a [test loader](/test/api_integration/apis/index.js). More about the existing FTR setup in the [contribution guide](/CONTRIBUTING.md#running-specific-kibana-tests)
```

Comment:
These relative links won't work in the docs system. Use the full github link (until that's an mdx file too!)